Refactor IO::Syscall as IO::Evented

ysbaddaden opened the pull request:
I started this refactor in my MT experiments, but I think it stands on its own:

`Evented` is closer to what the module is trying to achieve (abstract event loop details), and method naming follows that from `IO::Buffered` (e.g. `evented_read` instead of `read_syscall_helper`).

Refactors the `FileDescriptor` and `Socket` IO classes to move implementation details into `IO::Evented` to 1. reduce duplication, and 2. avoid direct accesses to internal details (such as `@writers` or `@read_event`).

ysbaddaden commented:
Huum, I think it should be `:nodoc:`. Maybe moved to `Crystal` namespace, along `Event` and `EventLoop`. Maybe even moved to `Crystal::System` since it's only used on POSIX platforms.

yxhuvud commented:
(quite OT, but related to the greater topic of evented IO on Linux) I wonder how much of this will change when https://lwn.net/Articles/776703/ is fully merged and accessible from userspace.

ysbaddaden commented:
Having them moved back into IO::Evented helped in my MT experiments, thought I moved most libevent responsibility into EventLoop, and use a single Event on Fiber directly, which holds the status (e.g. added, canceled). See https://github.com/ysbaddaden/crystal/commit/2ea421a2858db3398d1c6314b56031ed4137f79f

ysbaddaden commented:
@bcardiff it was already :)

bew commented:
It looks like everybody agree, let's get going! :rocket: 
![ok](https://user-images.githubusercontent.com/9730330/54446799-12321400-4748-11e9-8b1f-47e39e5b39ab.png)

straight-shoota commented:
Thank you @ysbaddaden!
